### PR TITLE
HIP buffer transfer - Eliminate cupy usage

### PR DESCRIPTION
**Pull request by swetha097**
Support for Hip buffer transfer by eliminate the usage of cupy

### CODE REVIEW

File: rocAL/rocAL/source/rali_api_data_transfer.cpp [hunk 1]
```diff
@@ -111,3 +111,48 @@ raliCopyToOutput(
...
+RaliStatus RALI_API_CALL
+raliCopyToHipOutputTensor32(RaliContext p_context,  void *out_ptr, RaliTensorLayout tensor_format, float multiplier0,
+                       float multiplier1, float multiplier2, float offset0, float offset1, float offset2,
+                       bool reverse_channels)
+{
+    auto context = static_cast<Context*>(p_context);
+    try
+    {
+        auto tensor_layout = (tensor_format == RALI_NHWC) ?  RaliTensorFormat::NHWC : RaliTensorFormat::NCHW;
+        context->master_graph->copy_hip_out_tensor(out_ptr, tensor_layout, multiplier0, multiplier1, multiplier2,
```
Comment (rrawther) [inline rocAL/rocAL/source/rali_api_data_transfer.cpp:123]: use the copy_out_tensor from rocAL with device pointer

File: rocAL/rocAL_pybind/amd/rali/pipeline.py [hunk 2]
```diff
@@ -240,22 +241,36 @@ def copyImage(self, array):
         b.raliCopyToOutput(
             self._handle, np.ascontiguousarray(out, dtype=array.dtype))
 
+    def copyToHipTensorNHWC(self, array, multiplier, offset, reverse_channels, tensor_dtype):
+        if tensor_dtype == types.FLOAT:
+            b.raliCopyToHipOutputTensor32(self._handle, ctypes.c_void_p(array.data_ptr()), types.NHWC,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rali/pipeline.py:246]: make changes as per the review comments for c++ api

File: rocAL/rocAL_pybind/rali_pybind.cpp [hunk 3]
```diff
@@ -35,6 +35,18 @@ namespace pybind11
 namespace rali{
     using namespace pybind11::literals; // NOLINT
     // PYBIND11_MODULE(rali_backend_impl, m) {
+        static void *ctypes_void_ptr(const py::object &object)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rali_pybind.cpp:38]: adjust spacing here

Comment (swetha097) [reply rocAL/rocAL_pybind/rali_pybind.cpp]: Done

File: rocAL/rocAL/include/rali_api_data_transfer.h [hunk 4]
```diff
@@ -43,4 +43,15 @@ extern "C"  RaliStatus   RALI_API_CALL raliCopyToOutputTensor16(RaliContext rali
                                                               float multiplier1, float multiplier2, float offset0,
                                                               float offset1, float offset2,
                                                               bool reverse_channels);
+
+extern "C" RaliStatus RALI_API_CALL raliCopyToHipOutputTensor32(RaliContext p_context, void *out_ptr, RaliTensorLayout tensor_format,
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/rali_api_data_transfer.h]: Done

File: rocAL/rocAL_pybind/amd/rali/pipeline.py (hunk 2 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rali/pipeline.py]: Done

**Comment by swetha097** [comment]
> reviewed with comments

@rrawther 
Addressed the review comments
Confirmed by running the host backend & hip backend trainings in the MI100 Machine

**Review by rrawther — APPROVED** [review]
@LakshmiKumar23 : please verify classification and SSD trainings with these changes

**Comment by kiritigowda** [comment]
@rrawther - CI pipeline enabled - http://math-ci.rocm.amd.com/blue/organizations/jenkins/compute-rocm-dkms-no-npi-hipclang%2FShort-GPUOpen%2FMIVisionX/detail/PR-765/1/pipeline

@LakshmiKumar23 let us know if this looks good with your test.

**Comment by LakshmiKumar23** [comment]
> @LakshmiKumar23 : please verify classification and SSD trainings with these changes

@rrawther Does this need to be run on any particular hardware specifically?

**Comment by kiritigowda** [comment]
@rrawther rerunning CI due to failure. @LakshmiKumar23 can you confirm this PR for merge.

**Comment by LakshmiKumar23** [comment]
do

> @rrawther rerunning CI due to failure. @LakshmiKumar23 can you confirm this PR for merge.

@kiritigowda I am trying to test it now

**Comment by kiritigowda** [comment]
@LakshmiKumar23 any updates on this?

**Review by rrawther — COMMENTED** [review]
@swetha097 : There are few codacy errors. Can you fix those so we can merge this PR.
Also it seems to have merge conflicts with tot

**Comment by kiritigowda** [comment]
> @kiritigowda : Can you merge this PR. Swetha has confirmed it is working

waiting for CI to finish -> http://math-ci.rocm.amd.com/blue/organizations/jenkins/compute-rocm-dkms-no-npi-hipclang%2FShort-GPUOpen%2FMIVisionX/detail/PR-765/4/pipeline

**Pull request merged**